packages/headless/src/app/engine-configuration-options.ts

```diff
@@ -0,0 +1,109 @@
+   */
+  accessToken: string;
+  /**
+   * A function that fetches a new access token. The function must return a Promise that resolves to a string (the new access token).
+   */
+  renewAccessToken?: () => Promise<string>;
+  /**
+   * Allows for augmenting a Platform request before it is sent.
+   * @param request Request to be augmented
+   * @param clientOrigin The origin of the client, can be "analyticsFetch", "analyticsBeacon" or "searchApiFetch"
+   *
+   * @returns Augmented request
+   */
+  preprocessRequest?: PreprocessRequest;
+  /**
+   * The Plaform URL to use. (e.g., https://platform.cloud.coveo.com)
+   * The platformUrl() helper method can be useful to know what url is available.
+   */
+  platformUrl?: string;
+  /**
+   * Allows configuring options related to analytics.
+   */
+  analytics?: {
+    /**
+     * Specifies if usage analytics tracking should be enabled.
+     *
+     * By default, all analytics events will be logged.
+     */
+    enabled?: boolean;
+    /**
+     * Origin level 2 is a usage analytics event metadata whose value should typically be the name/identifier of the tab from which the usage analytics event originates.
+     *
+     * When logging a Search usage analytics event, originLevel2 should always be set to the same value as the corresponding tab (parameter) Search API query parameter so Coveo Machine Learning models function properly, and usage analytics reports and dashboards are coherent.
+     *
+     * This value is optional, and will automatically try to resolve itself from the tab search parameter.
+     */
+    originLevel2?: string;
+    /**
+     * Origin level 3 is a usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface that’s making the request.
+     *
+     * When logging a Search usage analytics event, originLevel3 should always be set to the same value as the corresponding referrer Search API query parameter so usage analytics reports and dashboards are coherent.
+     *
+     * This value is optional, and will automatically try to resolve itself from the referrer search parameter.
+     */
+    originLevel3?: string;
+    /**
+     * analyticsClientMiddleware allows to hook into an analytics event payload before it is sent to the Coveo platform.
+     */
+    analyticsClientMiddleware?: AnalyticsClientSendEventHook;
+    /**
+     * Optional analytics runtime environment, this is needed for analytics to work correctly if you're running outside of a browser.
+     * See https://github.com/coveo/coveo.analytics.js for more info.
+     */
+    runtimeEnvironment?: IRuntimeEnvironment;
+  };
+}
+
+export const engineConfigurationOptionDefinitions: SchemaDefinition<EngineConfigurationOptions> = {
+  organizationId: new StringValue({
```

Comment:
Suggestion:

Since some of them are repeated, could help readability to extract them:

```
const requiredNonEmptyString = ()=> new StringValue({required: true, emptyAllowed: false})

const nonRequiredStringValue = ()=> new StringValue({required: false})

[...]

const engineConfigurationOptionDefinitions: SchemaDefinition<EngineConfigurationOptions> = {
   organizationId: requiredNonEmptyString(),
   accessToken: requiredNonEmptyString(),
   [...]
}
```

Reply:
Hmm, I tried this, but what I noticed is:
- The extracted function is used a maximum of twice, some are only used once.
- If I do not extract those that are used once, then there is a mix of levels of abstraction, where some options are calling a function, others are using the `new` keyword.

I feel it would more consistent to use one form for all the definitions, and extracting a combination that is at most used twice is not saving much.

Reply:
`requiredNonEmptyString` is a validation that is needed quite often. Doing a quick search, I can find at least 12 occurrences.
The funny thing is that this exact function has been implemented in both `analytics-utils.ts` and `validate-payload.ts` :)

I guess it would make sense to have Bueno expose shortcuts for those common use cases.

Reply:
I see, I was extracting the validation into a variable local to file. I will reuse the exported one and clear up the duplicated declaration in `analytics-utils`. Thanks,